torch/testing/_asserts.py

```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
+    """Parses the positional inputs by constructing :class:`_TensorPairs` from corresponding array-or-scalar-likes.
 
-    Regardless of the input types, the sequences are returned as :class:`dict`.
+    :class:`~collections.abc.Sequence`'s or :class:`~collections.abc.Mapping`'s are parsed elementwise. Parsing is
+    performed recursively and thus nested containers are supported. The hierarchy of the containers is preserved, but
+    sequences are returned as :class:`list` and mappings as :class:`dict`.
 
     Args:
-        actual: Actual mapping array-or-scalar-likes.
-        expected: Expected mapping array-or-scalar-likes.
+        actual (Any): Actual input.
+        expected (Any): Expected input.
 
     Returns:
-        (Optional[Exception], Optional[_ParsedInputs]): The two elements are orthogonal, i.e. if the first ``is None``
-            the second will not and vice versa. Returns a :class:`AssertionError` if the keys of :attr:`actual` and
-            :attr:`expected` do not match. Additionally, returns any exception from
-            :func:`_parse_array_or_scalar_like_pair`.
+        (Tuple[Optional[Exception], Optional[Union[_TensorPair, List, Dict]]]): The two elements are orthogonal, i.e.
+            if the first ``is None`` the second will be valid and vice versa. Returns an :class:`AssertionError` if the
+            length of two sequences or the keys of two mappings do not match. Additionally, returns any exception from
+            :func:`_to_tensor_pair`.
     """
     exc: Optional[Exception]
 
-    actual_keys = set(actual.keys())
-    expected_keys = set(expected.keys())
-    if actual_keys != expected_keys:
-        missing_keys = expected_keys - actual_keys
-        additional_keys = actual_keys - expected_keys
-        exc = AssertionError(
-            f"The keys of the mappings do not match:\n\n"
-            f"Missing keys in the actual mapping: {sorted(missing_keys)}\n"
-            f"Additional keys in the actual mapping: {sorted(additional_keys)}\n"
-        )
-        return exc, None
+    # We explicitly exclude str's here since they are self-referential and would cause an infinite recursion loop:
+    # "a" == "a"[0][0]...
+    if (
+        isinstance(actual, collections.abc.Sequence)
+        and not isinstance(actual, str)
+        and isinstance(expected, collections.abc.Sequence)
+        and not isinstance(expected, str)
+    ):
+        actual_len = len(actual)
+        expected_len = len(expected)
+        if actual_len != expected_len:
+            exc = AssertionError(f"The length of the sequences mismatch: {actual_len} != {expected_len}")
+            return exc, None
+
+        pair_list = []
+        for idx in range(actual_len):
+            exc, pair = _parse_inputs(actual[idx], expected[idx])
+            if exc:
+                exc = _amend_error_message(exc, f"{{}}\n\n{_SEQUENCE_MSG_FMTSTR.format(idx)}")
+                return exc, None
```

Comment:
I've inlined this from `_parse_sequences`. Although this makes this function harder to read, I wanted to avoid alternating recursive call between two functions.